SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLoadBalancingHandler.h

```diff
@@ -0,0 +1,170 @@
+#include "Schema/SpatialDebugging.h"
+#include "Utils/SpatialActorUtils.h"
+
+// Template class to handle LoadBalancing for a collection of actors.
+// Since different systems (NetDriver, ReplicationGraph) have different types for actor collection
+// and different ways to declare dependencies between actors, a CRTP pattern is used to have the
+// implementation class provide access to these information.
+// The methods to provide are :
+//  - GetActorsBeingReplicated() -> returns a range-for compatible iterator over AActor being replicated this frame
+//  - AddActorToReplicate(AActor* Actor) -> inform that an additional actor should be replicated this frame (a dependent actor)
+//  - RemoveAdditionalActor(AActor* Actor) -> inform that an actor we signaled as needing to replicate was encountered in the list of replicated actor, and does not need additional handling
+//  - GetDependentActors(AActor* Actor) -> returns a range-for compatible iterator over AActor that depends on the given actor (and should be migrated together)
+template <typename Implementation>
+class TSpatialLoadBalancingHandler
+{
+public:
+
+	TSpatialLoadBalancingHandler(USpatialNetDriver* InNetDriver)
+		: NetDriver(InNetDriver)
+	{
+
+	}
+
+	// Iterates over the list of actors to replicate, to check if they should migrate to another worker
+	// and collects additional actors to replicate if needed.
+	void HandleLoadBalancing()
+	{
+		TSet<FNetworkObjectInfo*> AdditionalObjectsToConsider;
+
+		for (AActor* Actor : static_cast<Implementation*>(this)->GetActorsBeingReplicated())
+		{
+			Worker_EntityId EntityId = NetDriver->PackageMap->GetEntityIdFromObject(Actor);
+			if (EntityId == SpatialConstants::INVALID_ENTITY_ID)
+			{
+				continue;
+			}
+
+			if (!Actor->HasAuthority())
+			{
+				continue;
+			}
+
+			if (SpatialGDK::SpatialDebugging* DebuggingInfo = NetDriver->StaticComponentView->GetComponentData<SpatialGDK::SpatialDebugging>(EntityId))
+			{
+				const bool bIsLocked = NetDriver->LockingPolicy->IsLocked(Actor);
+				if (DebuggingInfo->IsLocked != bIsLocked)
+				{
+					DebuggingInfo->IsLocked = bIsLocked;
+					FWorkerComponentUpdate DebuggingUpdate = DebuggingInfo->CreateSpatialDebuggingUpdate();
+					NetDriver->Connection->SendComponentUpdate(EntityId, &DebuggingUpdate);
+				}
+			}
+
+			// If this object is in the list of actors to migrate, we have already processed its hierarchy.
+			// Remove it from the additional actors to process, and continue.
+			if (ActorsToMigrate.Contains(Actor))
+			{
+				static_cast<Implementation*>(this)->RemoveAdditionalActor(Actor);
+				continue;
+			}
+
+			// TODO: the 'bWroteSomethingImportant' check causes problems for actors that need to transition in groups (ex. Character, PlayerController, PlayerState),
+			// so disabling it for now.  Figure out a way to deal with this to recover the perf lost by calling ShouldChangeAuthority() frequently. [UNR-2387]
+			if (NetDriver->LoadBalanceStrategy != nullptr &&
```

Comment:
Extract to a check, should always be set

Reply:
So this is impossible to mess up by the user ? Meaning this won't crash the editor if the user misconfigured it ?

Reply:
If it's unset, we default it, should be fine to check